src/api/telemetry/TelemetryCollection.js

```diff
@@ -0,0 +1,358 @@
+        this._unwatchTimeSystem();
+        if (this.unsubscribe) {
+            this.unsubscribe();
+        }
+    }
+
+    /**
+     * Sets up  the telemetry collection for historical requests,
+     * this uses the "standardizeRequestOptions" from Telemetry API
+     */
+    _initiateHistoricalRequests() {
+        this.openmct.telemetry.standardizeRequestOptions(this.options);
+        this.historicalProvider = this.openmct.telemetry.
+            findRequestProvider(this.domainObject, this.options);
+
+        this._requestHistoricalTelemetry();
+    }
+    /**
+     * If a historical provider exists, then historical requests will be made
+     */
+    async _requestHistoricalTelemetry() {
+        if (!this.historicalProvider) {
+            return;
+        }
+
+        let historicalData;
+
+        try {
+            historicalData = await this.historicalProvider.request(this.domainObject, this.options);
+        } catch (error) {
+            console.error('Error requesting telemetry data...');
+            throw new Error(error);
+        }
+
+        this._processNewTelemetry(historicalData);
+
+    }
+    /**
+     * This uses the built in subscription function from Telemetry API
+     */
+    _initiateSubscriptionTelemetry() {
+
+        if (this.unsubscribe) {
+            this.unsubscribe();
+        }
+
+        this.unsubscribe = this.openmct.telemetry
+            .subscribe(this.domainObject, (datum) => {
+                this._processNewTelemetry(datum);
+            }, this.options);
+    }
+
+    /**
+     * call back for telemetry provider to add more data as well as
+     * pass in the current state of the telemetry collection
+     * (which the telemetry collection will hold)
+     *
+     *
+     * @param  {Object[]} telemetryData - array of telemetry data objects
+     * @param  {*} [collectionState] - providers can pass a collectionState that
+     * will be used for tracking between collection and provider
+     */
+    _addPage(telemetryData, collectionState) {
+        this.collectionState = collectionState;
+        this._processNewTelemetry(telemetryData);
+    }
+
+    /**
+     * Filter any new telemetry (add/page, historical, subscription) based on
+     * time bounds
+     *
+     * @param  {(Object|Object[])} telemetryData - telemetry data object or
+     * array of telemetry data objects
+     */
+    _processNewTelemetry(telemetryData) {
+        let data = Array.isArray(telemetryData) ? telemetryData : [telemetryData];
+        let parsedValue;
+        let beforeStartOfBounds;
+        let afterEndOfBounds;
+        let added = [];
+
+        for (let datum of data) {
+            parsedValue = this.parseTime(datum);
+            beforeStartOfBounds = parsedValue < this.lastBounds.start;
+            afterEndOfBounds = parsedValue > this.lastBounds.end;
+
+            if (!afterEndOfBounds && !beforeStartOfBounds) {
+                if (!this.boundedTelemetry.includes(datum)) {
```

Comment:
Is this to avoid dupes? If so, `.includes()` won't work here, it just does reference equality.

Another thing I'm noticing is that it's not doing a sorted insertion. Unfortunately we can't assume that data will arrive in ascending order (or indeed in any order at all). The SortedTableRowCollection uses `_.sortedIndexBy` for this purpose which uses a binary search.

The `SortedTableRowCollection` also [handles deduplication](https://github.com/nasa/openmct/blob/master/src/plugins/telemetryTable/collections/SortedTableRowCollection.js#L90) using some performance optimizations.

Reply:
Ok, yeah, I'll update this to utilize what SortedTableRowCollection is currently handling.